extensions/battleroyale/standalone.mjs

```diff
@@ -0,0 +1,532 @@
+  }
+
+  /**
+   * Game interval
+   */
+  setGameInterval() {
+    alarm.recurring(START_GAME_INTERVAL, this.startGame.bind(this));
+  }
+
+  /**
+   * start Game
+   * @return {Boolean} success or not
+   */
+  async startGame() {
+    if (this.gameStarted) {
+      await this.resetGame();
+    }
+    if (this.participatePlayerIDs.size < 3) {
+      for (const playerID of this.participatePlayerIDs) {
+        await this.helper.callS2cAPI(playerID, 'notification', 'showNotification', 3000, '[battleroyale] player not enough to start (less than 3)');
+      }
+      return false;
+    }
+    for (const playerID of this.participatePlayerIDs) {
+      const playerloc = await this.helper.getPlayerLocation(playerID);
+      this.playerOldPosition.set(playerID, playerloc);
+      await this.helper.callS2cAPI(playerID, 'notification', 'showNotification', 3000, '[battleroyale] Game Start!');
+      await this.teleportPlayer(playerID);
+    }
+
+    this.gameStarted = true;
+    const updateMap = setInterval(async () => {
+      for (const mapName of this.arenaOfMaps.keys()) {
+        const needUpdate = this.updateEdge(
+            this.fires, this.arenaOfMaps.get(mapName), this.fireCounter,
+        );
+        if (!needUpdate) {
+          clearInterval(updateMap);
+          return;
+        }
+        await this.helper.broadcastCellSetUpdateToAllUser(
+            'update',
+            mapName,
+            CellSet.fromObject({
+              name: LAYER_FIRE.layerName,
+              cells: Array.from(this.fires),
+            }),
+        );
+        const players = this.helper.gameState.getPlayers();
+        players.forEach((player, playerID) => {
+          if (this.helper.gameMap.getCell(player.mapCoord, LAYER_FIRE.layerName)) {
+            this.killPlayer(playerID);
+          }
+        });
+        this.fireCounter++;
+        this.fires.clear();
+      }
+    }, MAP_UPDATE_PERIOD);
+    this.updateMapIntervalIDs.push(updateMap);
+
+    return true;
+  }
+
+  /**
+   * Given a Coordinate, return inside battleroyale arena or not
+   * @param {MapCoord} cellCoord - The cell position.
+   * @return {Boolean} inside or not
+   */
+  insideMap(cellCoord) {
+    for (const {cells} of this.arenaOfMaps.get(cellCoord.mapName)) {
+      for (const cell of cells) {
+        const width = cell.w ?? 1;
+        const height = cell.h ?? 1;
+        if (cell.x <= cellCoord.x && cellCoord.x < cell.x + width &&
+          cell.y <= cellCoord.y && cellCoord.y < cell.y + height) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Client tries to shoot bullet in front(depends on facing) of mapCoord.
+   * The server returns true on success.
+   * @param {Object} player - (TODO: needs docs)
+   * @param {MapCoord} mapCoord - The bullet position.
+   * @param {String} facing - The bullet facing.
+   * @return {Boolean} success or not
+   */
+  async c2s_attack(player, mapCoord, facing) {
```

Comment:
I think that this function only emits the bullet. I think the logic of bullet updating and collision detection should be separated into another function?

Reply:
If I understood correctly, the current layout of your function is like

```javascript
async c2s_attack(...) {
  // check if player can attack
  // create bullet
  // check if the bullet hits someone
  setInterval(() => {
    // update bullet
    // check if the bullet hits someone
  });
}
```

, that would have a lot of code duplicated.
Better, you can do this:

```javascript
Bullet.update = (firstTime) => {
  if (!firstTime) {
    // update bullet
  }
  // check if the bullet hits someone
}

async c2s_attack(...) {
  // check if player can attack
  // create bullet
  bullet.update();
  setInterval(bullet.update);
}
```